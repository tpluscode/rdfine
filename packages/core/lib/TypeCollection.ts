import { DatasetCore, Term } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import { RdfResource, ResourceIdentifier } from '../RdfResource'
import { rdf } from '@tpluscode/rdf-ns-builders'
import RDF from '@rdfjs/data-model'
import { onlyUnique } from './filter'
import * as compare from './compare'

function getNode(value: RdfResource | ResourceIdentifier | string): ResourceIdentifier {
  if (typeof value === 'string') {
    return RDF.namedNode(value)
  }

  if ('termType' in value) {
    return value
  }

  return value.id
}

export interface TypeCollection<D extends DatasetCore> extends Set<RdfResource<D>> {
  add(value: RdfResource<D> | ResourceIdentifier | string): this
  delete(value: RdfResource<D> | ResourceIdentifier | string): boolean
  has(value: RdfResource<D> | ResourceIdentifier | string): boolean
}

export default class <D extends DatasetCore> implements Set<RdfResource<D>> {
  private readonly __resource: RdfResource<D>
  private readonly __allGraphs: boolean
  private __graph: SingleContextClownface<Term, D>

  add(value: RdfResource<D> | ResourceIdentifier | string): this {
    this.__resource._selfGraph.addOut(rdf.type, getNode(value))
    return this
  }

  clear(): void {
    this.__graph.deleteOut(rdf.type)
  }

  delete(value: RdfResource<D> | ResourceIdentifier | string): boolean {
    const deletedQuads = this.__graph.dataset.match(this.__resource.id, rdf.type, getNode(value))

    for (const quad of deletedQuads) {
      this.__graph.dataset.delete(quad)
    }

    return deletedQuads.size > 0
  }

  forEach(callbackfn: (value: RdfResource<D>, value2: RdfResource<D>, set: Set<RdfResource<D>>) => void, thisArg?: unknown): void {
    for (const value of this.values()) {
      callbackfn.call(thisArg, value, value, this)
    }
  }

  has(value: RdfResource<D> | ResourceIdentifier | string): boolean {
    return this.__graph.has(rdf.type, getNode(value)).terms.length > 0
  }

  get size(): number {
    return this.__graph.out(rdf.type).terms.filter(onlyUnique(compare.terms)).length
  }

  [Symbol.iterator](): IterableIterator<RdfResource<D>> {
    return this.values()
  }

  entries(): IterableIterator<[RdfResource<D>, RdfResource<D>]> {
    return this.__values.map<[RdfResource<D>, RdfResource<D>]>(res => [res, res])[Symbol.iterator]()
  }

  keys(): IterableIterator<RdfResource<D>> {
    return this.values()
  }

  values(): IterableIterator<RdfResource<D>> {
    return this.__values[Symbol.iterator]()
  }

  [Symbol.toStringTag]: string;

  public constructor(resource: RdfResource<D>, allGraphs = false) {
    this.__resource = resource
    this.__allGraphs = allGraphs
    this.__graph = allGraphs ? cf({ dataset: resource._selfGraph.dataset, term: resource._selfGraph.term, graph: undefined }) : resource._selfGraph
  }

  private get __values() {
    const graphId = !this.__allGraphs ? this.__resource._graphId : null
    const typeQuads = this.__graph.dataset.match(this.__resource.id, rdf.type, null, graphId)

    return [...typeQuads]
      .map(quad => {
        return cf({
          dataset: this.__graph.dataset,
          term: quad.object,
          graph: quad.graph,
        })
      })
      .map(type => this.__resource._create<RdfResource<D>>(type))
      .filter(onlyUnique(compare.resources(false)))
  }
}
