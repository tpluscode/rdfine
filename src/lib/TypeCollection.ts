import { BlankNode, DatasetCore, NamedNode } from 'rdf-js'
import { RdfResource } from './RdfResource'
import { rdf } from './vocabs'
import { namedNode } from 'rdf-data-model'

function getNode(value: RdfResource | NamedNode | BlankNode | string): NamedNode | BlankNode {
  if (typeof value === 'string') {
    return namedNode(value)
  }

  if ('termType' in value) {
    return value
  }

  return value.id
}

export class TypeCollection<D extends DatasetCore> implements Set<RdfResource<D>> {
  private readonly __resource: RdfResource<D>

  add(value: RdfResource<D> | NamedNode | BlankNode | string): this {
    this.__resource._node.addOut(rdf.type, getNode(value))
    return this
  }

  clear(): void {
    this.__resource._node.deleteOut(rdf.type)
  }

  delete(value: RdfResource<D> | NamedNode | BlankNode | string): boolean {
    const deletedQuads = this.__resource._node.dataset.match(this.__resource.id, rdf.type, getNode(value))

    for (const quad of deletedQuads) {
      this.__resource._node.dataset.delete(quad)
    }

    return deletedQuads.size > 0
  }

  forEach(callbackfn: (value: RdfResource<D>, value2: RdfResource<D>, set: Set<RdfResource<D>>) => void, thisArg?: unknown): void {
    for (const value of this.values()) {
      callbackfn.call(thisArg, value, value, this)
    }
  }

  has(value: RdfResource<D> | NamedNode | BlankNode | string): boolean {
    return this.__resource._node.has(rdf.type, getNode(value)).terms.length > 0
  }

  get size(): number {
    return this.__resource._node.out(rdf.type).terms.length
  }

  [Symbol.iterator](): IterableIterator<RdfResource<D>> {
    return this.values()
  }

  entries(): IterableIterator<[RdfResource<D>, RdfResource<D>]> {
    return this.__values.map<[RdfResource<D>, RdfResource<D>]>(res => [res, res])[Symbol.iterator]()
  }

  keys(): IterableIterator<RdfResource<D>> {
    return this.values()
  }

  values(): IterableIterator<RdfResource<D>> {
    return this.__values[Symbol.iterator]()
  }

  [Symbol.toStringTag]: string;

  public constructor(resource: RdfResource<D>) {
    this.__resource = resource
  }

  private get __values() {
    return this.__resource._node
      .out(rdf.type)
      .map(type => {
        if (type.term.termType !== 'BlankNode' && type.term.termType !== 'NamedNode') {
          throw new Error('RDF Type must be a resource')
        }

        return this.__resource._create<RdfResource<D>>(type as any)
      })
  }
}
