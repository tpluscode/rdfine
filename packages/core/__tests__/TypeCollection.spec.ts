import $rdf from 'rdf-ext'
import cf, { GraphPointer } from 'clownface'
import TypeCollection from '../lib/TypeCollection'
import RdfResourceImpl, { RdfResource, ResourceIdentifier } from '../RdfResource'
import { parse, ex } from './_helpers'
import { rdf } from '@tpluscode/rdf-ns-builders'

const nullResource = {} as RdfResource

describe('TypeCollection', () => {
  let node: GraphPointer<ResourceIdentifier>

  describe('size', () => {
    it('returns 0 when no types', () => {
      // given
      const dataset = $rdf.dataset()
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.size).toEqual(0)
    })

    it('counts types', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type1, ex:Type2, ex:Type3, ex:Type4 .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.size).toEqual(4)
    })

    it('counts unique types', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type1, ex:Type2, ex:Type1, ex:Type2 .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.size).toEqual(2)
    })
  })

  describe('has', () => {
    it('returns true for type found by string', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.has('http://example.com/Type')).toBe(true)
    })

    it('returns true for type found by exact node', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.has(ex.Type)).toBe(true)
    })

    it('returns true for type found by resource', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const type: RdfResource = {
        ...nullResource,
        id: ex.Type,
      }

      // then
      expect(tc.has(type)).toBe(true)
    })

    it('returns false for type not found by string', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.has('http://example.com/Type2')).toBe(false)
    })

    it('returns false for type not found by exact node', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))

      // then
      expect(tc.has(ex.Type2)).toBe(false)
    })

    it('returns false for type not found by resource', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const type = {
        ...nullResource,
        id: ex.Type2,
      }

      // then
      expect(tc.has(type)).toBe(false)
    })
  })

  describe('clear', () => {
    it('removes all resource\'s triples', async () => {
      // given
      const dataset = await parse(`
      @prefix ex: <${ex().value}> .
      
      ex:res a ex:Type1, ex:Type2, ex:Type3, ex:Type4 .
      ex:res2 a ex:Type1 .
    `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      tc.clear()

      // then
      expect(dataset.size).toEqual(1)
      expect(dataset.toCanonical()).toMatchSnapshot()
    })
  })

  describe('add', () => {
    it('modifies the dataset when adding named node', () => {
      // given
      const dataset = $rdf.dataset()
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      tc.add(ex.Type)

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })

    it('modifies the dataset when adding resource', () => {
      // given
      const dataset = $rdf.dataset()
      node = cf({
        dataset,
      }).namedNode(ex.res)
      const newType = {
        ...nullResource,
        id: ex.Type,
      }

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      tc.add(newType)

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })

    it('does not modify the dataset when type already exists on resource', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)
      const newType = {
        ...nullResource,
        id: ex.Type,
      }

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      tc.add(newType)

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })
  })

  describe('delete', () => {
    it('removes the correct triple', async () => {
      // given
      const dataset = await parse(`
      @prefix ex: <${ex().value}> .
      
      ex:res a ex:Type1, ex:Type2, ex:Type3, ex:Type4 .
      ex:res2 a ex:Type1 .
    `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      tc.delete(ex.Type2)

      // then
      expect(dataset.size).toEqual(4)
      expect(dataset.toCanonical()).toMatchSnapshot()
    })

    it('returns true if deleted triples', async () => {
      // given
      const dataset = await parse(`
      @prefix ex: <${ex().value}> .
      
      ex:res a ex:Type1 .
    `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const deleted = tc.delete(ex.Type1)

      // then
      expect(deleted).toEqual(true)
    })

    it('returns false if deleted nothing', async () => {
      // given
      const dataset = await parse(`
      @prefix ex: <${ex().value}> .
      
      ex:res a ex:Type1 .
    `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const deleted = tc.delete(ex.Type2)

      // then
      expect(deleted).toEqual(false)
    })
  })

  describe('forEach', () => {
    it('iterates with thisArg and correct callback params', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const thisArg = {}

      // then
      tc.forEach(function (this: any, value, value2, set) {
        expect(ex.res.equals(value.id))
        expect(ex.res.equals(value2.id))
        expect(set).toBe(tc)
        expect(this).toBe(thisArg)
      }, thisArg)
    })
  })

  describe('entries', () => {
    it('implements the set interface', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type .
      `)
      node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new TypeCollection(new RdfResourceImpl(node))
      const entries = tc.entries()

      // then
      for (const [one, two] of entries) {
        expect(ex.Type.equals(one.id))
        expect(ex.Type.equals(two.id))
      }
    })
  })
})

describe('TypeCollection', () => {
  let node: GraphPointer<ResourceIdentifier>

  describe('allGraphs = true', () => {
    describe('size', () => {
      it('counts unique types from all graphs', () => {
        // given
        const dataset = $rdf.dataset()
        cf({ dataset, graph: ex.G1 })
          .namedNode(ex.res)
          .addOut(rdf.type, [ex.Type1, ex.Type2])
        cf({ dataset, graph: ex.G2 })
          .namedNode(ex.res)
          .addOut(rdf.type, [ex.Type1, ex.Type2])
        node = cf({
          dataset,
        }).namedNode(ex.res)

        // when
        const tc = new TypeCollection(new RdfResourceImpl(node), true)

        // then
        expect(tc.size).toEqual(2)
      })
    })

    describe('values', () => {
      it('returns unique types from all graphs', () => {
        // given
        const dataset = $rdf.dataset()
        cf({ dataset, graph: ex.G1 })
          .namedNode(ex.res)
          .addOut(rdf.type, [ex.Type1, ex.Type2])
        cf({ dataset, graph: ex.G2 })
          .namedNode(ex.res)
          .addOut(rdf.type, [ex.Type1, ex.Type2])
        node = cf({
          dataset,
        }).namedNode(ex.res)

        // when
        const tc = new TypeCollection(new RdfResourceImpl(node), true)
        const values = [...tc.values()]

        // then
        expect(values).toHaveLength(2)
        expect(values.map(r => r.id)).toEqual(
          expect.arrayContaining([ex.Type1, ex.Type2]),
        )
      })
    })
  })
})
