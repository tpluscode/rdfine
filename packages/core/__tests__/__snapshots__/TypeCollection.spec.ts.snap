// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeCollection add does not modify the dataset when type already exists on resource 1`] = `
"<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type> .
"
`;

exports[`TypeCollection add modifies the dataset when adding named node 1`] = `
"<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type> .
"
`;

exports[`TypeCollection add modifies the dataset when adding resource 1`] = `
"<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type> .
"
`;

exports[`TypeCollection clear removes all resource's triples 1`] = `
"<http://example.com/res2> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type1> .
"
`;

exports[`TypeCollection delete removes the correct triple 1`] = `
"<http://example.com/res2> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type1> .
<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type1> .
<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type3> .
<http://example.com/res> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Type4> .
"
`;
